docs/CppInteroperability/Roadmap.md

```diff
@@ -0,0 +1,141 @@
+
+As per our goal’s specification, this method of importing reference types allows C++ interoperability to have a clear, native mapping for a common C++ API pattern that builds a more general purpose solution off of the pre-established reference type bridging from Objective-C. Additionally, Swift preserves the same safety properties as C++ while providing an even safer option for the common pattern where reference types have a retain and release operation.
+
+### Value types
+
+As mentioned above, swift types fall into two categories: value types and reference types. Value types are relatively simple in Swift, C, and Objective-C, however, the same is not true for C++ where value types may have complex lifetime operations that are used to express non-native concepts (to Swift) such as value types that own memory. This case of owned value types is specifically difficult to map on to Swift as it is novel to the language and does not always work well with the existing model for Swift value types that are trivially copyable.  
+
+#### Trivial types
+
+This roadmap will refer to trivial value types that do not hold pointers as “trivial types.” These types include primitive types such as integers and types which are composed of other trivial types. Trivial types are “owned” types that provide trivial lifetime operations: a copy is a copy of their bits and a destroy is a no-op. Trivial types have roughly the same mapping throughout Swift, C, Objective-C, and C++ making them trivial to import. Trivial types, their instances, methods on trivial types, and other APIs that use trivial types are generally considered to be safe and usable.
+
+**Pointer types**
+
+This roadmap will refer to trivial value types that hold pointers as “pointer types.” These types include pointers themselves and types which are composed of any pointer types (potentially including other types as well). The pointers held by pointer types refer to memory that is *not owned* by the pointer type (making pointer types a “view” or “projection” into memory). While pointer types are very similar to trivial types with respect to their lifetime operations and the fact that they map similarly in these four language, they differ in the fact that while they themselves are not inherently unsafe, they may be used in unsafe APIs (discussed later).
+
+#### Owned types
+
+This category of types refers to types that own memory. These types are pointer types where the type *does own* the memory that their pointers point to. C++ often uses copy constructors and destructors to manage the lifetime of owned types. Therefore, the Swift compiler should assume that pointer types with custom copy constructor and destructors own their memory (and should provide annotations to override this assumption).
```

Comment:
I am left wondering what custom means here.  Custom as in there is one defined or custom as in an explicit one provided?

Reply:
Types with non-trivial (user provided) copy constructors as defined by the C++ standard, or types which implicitly have these non-trivial copy constructors because they are composed of a type which has a user provided copy constructor. Does that make sense? I think I will just define "non-trivial copy constructor" to mean this in the definitions and then use that throughout the document. It's unfortunate to overlay the word trivial yet again but...

Reply:
So to clarify, all types with "custom copy constructors" are assumed to be owned types right? User provides annotations if they have to exclude some type from being an owned type. 

This annotation must be used only if the type behaves like a completely unowned type  e.g. types like `StringRef`, `ArrayRef` etc. 

I am wondering if this is a stronger property (i.e. more restrictive) than what is needed. Wouldn't it be sufficient if  the type annotation conveys "I don't have any method that returns a pointer to my storage?".

Reply:
I think you probably do need to talk here about how types having non-trivial value operations doesn't necessarily mean that they're self-contained value types (e.g. `std::vector<char*>`).  Generally, it's good to work out a little bit of the detail here; you want to show you've thought it through and have an idea of where the complexity is going to lie.